File: apps/api/schemas/costpulse.py
"""Pydantic schemas for CostPulse module."""

from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from enum import StrEnum
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

# ---------- Enums ----------


class MaterialCategory(StrEnum):
    concrete = "concrete"
    steel = "steel"
    finishing = "finishing"
    mep = "mep"
    timber = "timber"
    masonry = "masonry"
    other = "other"


class EstimateStatus(StrEnum):
    draft = "draft"
    approved = "approved"
    superseded = "superseded"


class EstimateConfidence(StrEnum):
    rough_order = "rough_order"
    preliminary = "preliminary"
    detailed = "detailed"


class EstimateMethod(StrEnum):
    ai_generated = "ai_generated"
    manual = "manual"
    imported = "imported"


class RfqStatus(StrEnum):
    draft = "draft"
    sent = "sent"
    responding = "responding"
    closed = "closed"


class PriceSource(StrEnum):
    government = "government"
    supplier = "supplier"
    crowdsource = "crowdsource"


class BoqItemSource(StrEnum):
    ai_extracted = "ai_extracted"
    manual = "manual"
    price_db = "price_db"


# ---------- Prices ----------


class MaterialPriceOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    material_code: str
    name: str
    category: MaterialCategory | None = None
    unit: str
    price_vnd: Decimal
    price_usd: Decimal | None = None
    province: str | None = None
    source: PriceSource | None = None
    effective_date: date
    expires_date: date | None = None
    supplier_id: UUID | None = None


class MaterialPriceQuery(BaseModel):
    q: str | None = Field(default=None, description="Full-text name/code search")
    material_code: str | None = None
    category: MaterialCategory | None = None
    province: str | None = None
    as_of: date | None = Field(default=None, description="Effective as of this date")
    limit: int = Field(default=50, ge=1, le=200)
    offset: int = Field(default=0, ge=0)


class PriceHistoryPoint(BaseModel):
    effective_date: date
    price_vnd: Decimal
    province: str | None = None
    source: PriceSource | None = None


class PriceHistoryResponse(BaseModel):
    material_code: str
    name: str
    unit: str
    points: list[PriceHistoryPoint]
    pct_change_30d: float | None = None
    pct_change_1y: float | None = None


# ---------- Estimates ----------


class EstimateFromBriefRequest(BaseModel):
    project_id: UUID | None = None
    name: str = Field(min_length=1, max_length=200)
    project_type: str = Field(description="e.g. residential, commercial, villa, factory")
    area_sqm: float = Field(gt=0)
    floors: int = Field(ge=1)
    province: str = Field(description="e.g. Hanoi, HCMC")
    quality_tier: Literal["economy", "standard", "premium"] = "standard"
    structure_type: Literal["reinforced_concrete", "steel", "mixed"] = "reinforced_concrete"
    notes: str | None = None


class EstimateFromDrawingsRequest(BaseModel):
    project_id: UUID | None = None
    name: str = Field(min_length=1, max_length=200)
    drawing_file_ids: list[UUID] = Field(min_length=1, max_length=50)
    province: str
    include_contingency_pct: float = Field(default=10.0, ge=0, le=50)


class BoqItemIn(BaseModel):
    id: UUID | None = None
    parent_id: UUID | None = None
    sort_order: int = 0
    code: str | None = None
    description: str = Field(min_length=1)
    unit: str | None = None
    quantity: Decimal | None = None
    unit_price_vnd: Decimal | None = None
    total_price_vnd: Decimal | None = None
    material_code: str | None = None
    source: BoqItemSource | None = None
    notes: str | None = None


class BoqItemOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    estimate_id: UUID
    parent_id: UUID | None = None
    sort_order: int
    code: str | None = None
    description: str
    unit: str | None = None
    quantity: Decimal | None = None
    unit_price_vnd: Decimal | None = None
    total_price_vnd: Decimal | None = None
    material_code: str | None = None
    source: BoqItemSource | None = None
    notes: str | None = None


class EstimateCreate(BaseModel):
    project_id: UUID | None = None
    name: str = Field(min_length=1, max_length=200)
    method: EstimateMethod = EstimateMethod.manual
    confidence: EstimateConfidence = EstimateConfidence.preliminary
    items: list[BoqItemIn] = Field(default_factory=list)


class EstimateSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    project_id: UUID | None = None
    name: str
    version: int
    status: EstimateStatus
    total_vnd: int | None = None
    confidence: EstimateConfidence | None = None
    method: EstimateMethod | None = None
    created_by: UUID | None = None
    approved_by: UUID | None = None
    created_at: datetime


class EstimateDetail(EstimateSummary):
    items: list[BoqItemOut] = Field(default_factory=list)


class UpdateBoqRequest(BaseModel):
    items: list[BoqItemIn]
    recompute_totals: bool = True


# ---------- Version diff ----------


class BoqDiffRow(BaseModel):
    """One line in a BOQ-version diff.

    `kind` is the verb the UI renders:
      * `added`   — present in `to`, not in `from`.
      * `removed` — present in `from`, not in `to`.
      * `changed` — present in both, but `(qty, unit_price, description)`
                    differs. The "from" + "to" snapshots are filled so
                    the UI can show side-by-side cells.
      * `unchanged` — same on both sides. Excluded from the response by
                      default; a `?include_unchanged=true` query param
                      would surface them, but the typical buyer view is
                      "what actually changed" so we keep the payload
                      lean.

    Match key: `material_code` first (canonical), falling back to a
    fold of `(code, description)` so a row a buyer typed by hand can
    still pair across versions even without a material catalogue match.
    """

    kind: Literal["added", "removed", "changed"]
    material_code: str | None = None
    code: str | None = None
    description: str
    """Always set — for added rows it's `to.description`, for removed
    rows `from.description`, for changed rows `to.description`. The UI
    displays this as the row label."""

    from_qty: Decimal | None = None
    to_qty: Decimal | None = None
    from_unit_price_vnd: Decimal | None = None
    to_unit_price_vnd: Decimal | None = None
    from_total_price_vnd: Decimal | None = None
    to_total_price_vnd: Decimal | None = None
    from_unit: str | None = None
    to_unit: str | None = None


class EstimateDiff(BaseModel):
    """`GET /estimates/{a_id}/diff?to={b_id}` payload."""

    from_version: int
    to_version: int
    from_total_vnd: int | None = None
    to_total_vnd: int | None = None
    rows: list[BoqDiffRow]
    """Sorted: added first, then changed, then removed; within each
    bucket by description so the diff is stable across reruns."""


# ---------- Suppliers ----------


class SupplierOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    organization_id: UUID | None = None
    name: str
    categories: list[str] = Field(default_factory=list)
    provinces: list[str] = Field(default_factory=list)
    contact: dict[str, Any] = Field(default_factory=dict)
    verified: bool
    rating: Decimal | None = None
    created_at: datetime


class SupplierCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    categories: list[str] = Field(default_factory=list)
    provinces: list[str] = Field(default_factory=list)
    contact: dict[str, Any] = Field(default_factory=dict)


# ---------- RFQ ----------


class RfqCreate(BaseModel):
    project_id: UUID | None = None
    estimate_id: UUID | None = None
    supplier_ids: list[UUID] = Field(min_length=1)
    deadline: date | None = None
    message: str | None = None
    material_codes: list[str] = Field(default_factory=list)


class RfqOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    project_id: UUID | None = None
    estimate_id: UUID | None = None
    status: RfqStatus
    sent_to: list[UUID] = Field(default_factory=list)
    responses: list[dict[str, Any]] = Field(default_factory=list)
    deadline: date | None = None
    # Buyer's accepted-quote columns from migration 0024_rfq_acceptance.
    # The frontend QuoteComparisonTable matches `accepted_supplier_id`
    # against each column to render the "✓ Accepted" badge.
    accepted_supplier_id: UUID | None = None
    accepted_at: datetime | None = None
    created_at: datetime


# ---------- Analytics ----------


class CostBenchmarkQuery(BaseModel):
    project_type: str | None = None
    province: str | None = None
    quality_tier: Literal["economy", "standard", "premium"] | None = None


class CostBenchmarkBucket(BaseModel):
    project_type: str
    province: str | None = None
    quality_tier: str | None = None
    cost_per_sqm_vnd_p25: int
    cost_per_sqm_vnd_median: int
    cost_per_sqm_vnd_p75: int
    sample_size: int


class CostBenchmarkResponse(BaseModel):
    buckets: list[CostBenchmarkBucket]


# ---------- AI pipeline output ----------


class AiEstimateResult(BaseModel):
    estimate_id: UUID
    total_vnd: int
    confidence: EstimateConfidence
    items: list[BoqItemOut]
    warnings: list[str] = Field(default_factory=list)
    missing_price_codes: list[str] = Field(default_factory=list)

File: packages/types/costpulse.ts
import type { ISODate, UUID } from "./envelope";

export type MaterialCategory = "concrete" | "steel" | "finishing" | "mep" | "timber" | "masonry" | "other";
export type EstimateStatus = "draft" | "approved" | "superseded";
export type EstimateConfidence = "rough_order" | "preliminary" | "detailed";
export type EstimateMethod = "ai_generated" | "manual" | "imported";
export type RfqStatus = "draft" | "sent" | "responded" | "responding" | "closed";

/** Supplier-side state for one (rfq, supplier) slot. Lives inside `Rfq.responses[]`. */
export type RfqResponseStatus =
  | "dispatched"  // email out, no quote yet
  | "bounced"     // email send failed
  | "skipped"     // no contact email / supplier hidden / not yet dispatched
  | "responded";  // supplier submitted via the public portal
export type PriceSource = "government" | "supplier" | "crowdsource";
export type BoqItemSource = "ai_extracted" | "manual" | "price_db";
export type QualityTier = "economy" | "standard" | "premium";

export interface MaterialPrice {
  id: UUID;
  material_code: string;
  name: string;
  category: MaterialCategory | null;
  unit: string;
  price_vnd: string;
  price_usd: string | null;
  province: string | null;
  source: PriceSource | null;
  effective_date: ISODate;
  expires_date: ISODate | null;
  supplier_id: UUID | null;
}

export interface PriceHistoryPoint {
  effective_date: ISODate;
  price_vnd: string;
  province: string | null;
  source: PriceSource | null;
}

export interface PriceHistory {
  material_code: string;
  name: string;
  unit: string;
  points: PriceHistoryPoint[];
  pct_change_30d: number | null;
  pct_change_1y: number | null;
}

export interface BoqItem {
  id: UUID;
  estimate_id: UUID;
  parent_id: UUID | null;
  sort_order: number;
  code: string | null;
  description: string;
  unit: string | null;
  quantity: string | null;
  unit_price_vnd: string | null;
  total_price_vnd: string | null;
  material_code: string | null;
  source: BoqItemSource | null;
  notes: string | null;
}

export interface BoqItemInput {
  id?: UUID | null;
  parent_id?: UUID | null;
  sort_order?: number;
  code?: string | null;
  description: string;
  unit?: string | null;
  quantity?: string | number | null;
  unit_price_vnd?: string | number | null;
  total_price_vnd?: string | number | null;
  material_code?: string | null;
  source?: BoqItemSource | null;
  notes?: string | null;
}

export interface EstimateSummary {
  id: UUID;
  project_id: UUID | null;
  name: string;
  version: number;
  status: EstimateStatus;
  total_vnd: number | null;
  confidence: EstimateConfidence | null;
  method: EstimateMethod | null;
  created_by: UUID | null;
  approved_by: UUID | null;
  created_at: string;
}

export interface EstimateDetail extends EstimateSummary {
  items: BoqItem[];
}

export interface EstimateFromBriefInput {
  project_id?: UUID | null;
  name: string;
  project_type: string;
  area_sqm: number;
  floors: number;
  province: string;
  quality_tier: QualityTier;
  structure_type: "reinforced_concrete" | "steel" | "mixed";
  notes?: string | null;
}

export interface EstimateFromDrawingsInput {
  project_id?: UUID | null;
  name: string;
  drawing_file_ids: UUID[];
  province: string;
  include_contingency_pct?: number;
}

export interface Supplier {
  id: UUID;
  organization_id: UUID | null;
  name: string;
  categories: string[];
  provinces: string[];
  contact: Record<string, unknown>;
  verified: boolean;
  rating: string | null;
  created_at: string;
}

/**
 * One quote line as submitted by a supplier through the public portal.
 *
 * Mirrors `apps/api/schemas/public_rfq.py::PublicRfqQuoteLine`. Numeric
 * fields land here as strings because the buyer-side surface ingests
 * them via Pydantic Decimal serialisation, which is JSON-serialised as
 * strings to avoid float drift.
 */
export interface RfqQuoteLine {
  material_code: string | null;
  description: string;
  quantity: number | null;
  unit: string | null;
  unit_price_vnd: string | null;
}

export interface RfqQuote {
  total_vnd: string | null;
  lead_time_days: number | null;
  valid_until: ISODate | null;
  notes: string | null;
  line_items: RfqQuoteLine[];
}

/** One per-supplier slot inside `Rfq.responses[]`. */
export interface RfqResponseEntry {
  supplier_id: UUID;
  status: RfqResponseStatus;
  /** Set when the dispatcher email-out happened. */
  dispatched_at?: string | null;
  /** Set when the supplier submitted through the public portal. */
  responded_at?: string | null;
  /** Email-transport bookkeeping written by `services.rfq_dispatch`. */
  delivery?: {
    to?: string;
    subject?: string;
    delivered: boolean;
    reason?: string | null;
  } | null;
  /** Null until the supplier submits a quote. */
  quote: RfqQuote | null;
}

export interface Rfq {
  id: UUID;
  project_id: UUID | null;
  estimate_id: UUID | null;
  status: RfqStatus;
  sent_to: UUID[];
  responses: RfqResponseEntry[];
  deadline: ISODate | null;
  /**
   * Buyer's accepted-quote columns from migration 0024_rfq_acceptance.
   * Set when the buyer picks a winner; null until then. Drives the
   * "✓ Accepted" badge in `QuoteComparisonTable`.
   */
  accepted_supplier_id: UUID | null;
  accepted_at: string | null;
  created_at: string;
}

export interface RfqInput {
  project_id?: UUID | null;
  estimate_id?: UUID | null;
  supplier_ids: UUID[];
  deadline?: ISODate | null;
  message?: string | null;
  material_codes?: string[];
}

export interface CostBenchmarkBucket {
  project_type: string;
  province: string | null;
  quality_tier: string | null;
  cost_per_sqm_vnd_p25: number;
  cost_per_sqm_vnd_median: number;
  cost_per_sqm_vnd_p75: number;
  sample_size: number;
}

export interface AiEstimateResult {
  estimate_id: UUID;
  total_vnd: number;
  confidence: EstimateConfidence;
  items: BoqItem[];
  warnings: string[];
  missing_price_codes: string[];
}
